port/raspberrypi/rp2xxx/src/hal/pins.zig

```diff
@@ -262,82 +460,249 @@ pub const Function = enum {
-    single(27), // ADC1
-    single(28), // ADC2
-    single(29), // ADC3
-};
+fn none() PinFlags {
+    return std.mem.zeroes(PinFlags);
+}
+
+const function_table = if (chip == .RP2040)
+    [@typeInfo(Function).@"enum".fields.len]PinFlags{
+        all(), // SIO
+        all(), // PIO0
+        all(), // PIO1
+        none(), // PIO2
+        list(&.{ 0, 4, 16, 20 }), // SPI0_RX
+        list(&.{ 1, 5, 17, 21 }), // SPI0_CSn
+        list(&.{ 2, 6, 18, 22 }), // SPI0_SCK
+        list(&.{ 3, 7, 19, 23 }), // SPI0_TX
+        list(&.{ 8, 12, 24, 28 }), // SPI1_RX
+        list(&.{ 9, 13, 25, 29 }), // SPI1_CSn
+        list(&.{ 10, 14, 26, 30 }), // SPI1_SCK
+        list(&.{ 11, 15, 27, 31 }), // SPI1_TX
+        list(&.{ 0, 12, 16, 28 }), // UART0_TX
+        list(&.{ 1, 13, 17 }), // UART0_RX
+        list(&.{ 2, 14, 18 }), // UART0_CTS
+        list(&.{ 3, 15, 19 }), // UART0_RTS
+        none(), // UART0_ALT_TX
+        none(), // UART0_ALT_RX
+        list(&.{ 4, 8, 20, 24 }), // UART1_TX
+        list(&.{ 5, 9, 21, 25 }), // UART1_RX
+        list(&.{ 6, 10, 22, 26 }), // UART1_CTS
+        list(&.{ 7, 11, 23, 27 }), // UART1_RTS
+        none(), // UART1_ALT_TX
+        none(), // UART1_ALT_RX
+        list(&.{ 0, 4, 8, 12, 16, 20, 24, 28 }), // I2C0_SDA
+        list(&.{ 1, 5, 9, 13, 17, 21, 25, 29 }), // I2C0_SCL
+        list(&.{ 2, 6, 10, 14, 18, 22, 26 }), // I2C1_SDA
+        list(&.{ 3, 7, 11, 15, 19, 23, 27 }), // I2C1_SCL
+        list(&.{ 0, 16 }), // PWM0_A
+        list(&.{ 1, 17 }), // PWM0_B
+        list(&.{ 2, 18 }), // PWM1_A
+        list(&.{ 3, 19 }), // PWM1_B
+        list(&.{ 4, 20 }), // PWM2_A
+        list(&.{ 5, 21 }), // PWM2_B
+        list(&.{ 6, 22 }), // PWM3_A
+        list(&.{ 7, 23 }), // PWM3_B
+        list(&.{ 8, 24 }), // PWM4_A
+        list(&.{ 9, 25 }), // PWM4_B
+        list(&.{ 10, 26 }), // PWM5_A
+        list(&.{ 11, 27 }), // PWM5_B
+        list(&.{ 12, 28 }), // PWM6_A
+        list(&.{ 13, 29 }), // PWM6_B
+        single(14), // PWM7_A
+        single(15), // PWM7_B
+        none(), // PWM8_A
+        none(), // PWM8_B
+        none(), // PWM9_A
+        none(), // PWM9_B
+        none(), // PWM10_A
+        none(), // PWM10_B
+        none(), // PWM11_A
+        none(), // PWM11_B
+        single(20), // CLOCK_GPIN0
+        single(22), // CLOCK_GPIN1
+        single(21), // CLOCK_GPOUT0
+        single(23), // CLOCK_GPOUT1
+        single(24), // CLOCK_GPOUT2
+        single(25), // CLOCK_GPOUT3
+        list(&.{ 0, 3, 6, 9, 12, 15, 18, 21, 24, 27 }), // USB_OVCUR_DET
+        list(&.{ 1, 4, 7, 10, 13, 16, 19, 22, 25, 28 }), // USB_VBUS_DET
+        list(&.{ 2, 5, 8, 11, 14, 17, 20, 23, 26, 29 }), // USB_VBUS_EN
+        single(26), // ADC0
+        single(27), // ADC1
+        single(28), // ADC2
+        single(29), // ADC3
+        none(), // ADC4
+        none(), // ADC5
+        none(), // ADC6
+        none(), // ADC7
+        none(), // QMI_CS1
+        none(), // HSTX
```

Comment:
doesn't the normal 2350 package still have HSTX on GPIO 12-19?

Reply:
It is on the RP2350A (see line 704) -- You're looking at the RP2040 function table.  

There are three versions of the function table.  In order RP2040, RP2350B and RP2350A.  I built separate tables for RP2040 and RP2350A so that trying to configure  nonexistent pins fails with the "pin cannot be configured for ..." message.

Reply:
Oops yeah I misread the part we were in :)